File: prozess_synchronisation/loesung/06/schlafender_friseur_semaphore.py
import threading
import time

# startet bei 0, da noch kein Kunde bei Ladenöffnung anwesend
semaphore = threading.Semaphore(0)

def barber():
    time.sleep(0.2) # initiale Verzögerung (öffnet den Laden)
    while True:
        semaphore.acquire()
        print("[Friseur] schneidet Haare ...")
    

def customer(counter):
    print(f"\t[Kunde #{counter}] betritt den Saloon.")
    print(f"\t[Kunde #{counter}] informiert Friseur.")
    semaphore.release()

# Threads starten
threading.Thread(target=barber).start()
for i in range(10):
    threading.Thread(target=customer, args=(i+1,)).start()
    time.sleep(0.5)

File: prozess_synchronisation/loesung/06/schlafender_friseur_semaphore.cpp
#include <iostream>
#include <thread>
#include <mutex>
#include <condition_variable>
#include <vector>
#include <chrono>

using namespace std;

class Semaphore
{
    mutex mtx;
    condition_variable cv;
    int count;

public:
    Semaphore(int c) : count(c) {}

    void acquire()
    {
        unique_lock<std::mutex> lock(mtx);
        cv.wait(lock, [&]
                { return count > 0; }); // warten, bis verfügbar
        --count;
    }

    void release()
    {
        std::unique_lock<std::mutex> lock(mtx);
        ++count;
        cv.notify_one();
    }
};

// startet bei 0, da noch kein Kunde bei Ladenöffnung anwesend
Semaphore sem(0);

void barber()
{
    this_thread::sleep_for(chrono::milliseconds(200));
    while (true)
    {
        cout << "[Friseur] schneidet Haare ..." << endl;
        sem.acquire();
    }
}

void customer(int counter)
{
    cout << "\t[Kunde #" << counter << "] betritt den Saloon." << endl;
    cout << "\t[Kunde #" << counter << "] informiert den Friseur." << endl;
    sem.release();
}

int main()
{
    thread barber_thread(barber);

    for (int i = 0; i < 10; i++)
    {
        thread customer_thread(customer, i + 1);
        customer_thread.join();
        this_thread::sleep_for(chrono::milliseconds(500));
    }

    barber_thread.detach();

    return 0;
}
